compiler/next/lib/resolution/scope-queries.cpp

```diff
@@ -0,0 +1,529 @@
+}
+
+static bool createsScope(asttags::ASTTag tag) {
+  return Builder::astTagIndicatesNewIdScope(tag)
+         || asttags::isSimpleBlockLike(tag)
+         || asttags::isLoop(tag)
+         || asttags::isCobegin(tag)
+         || asttags::isConditional(tag)
+         //|| asttags::isSelect(tag)
+         //|| asttags::isTryCatch(tag)
+         ;
+}
+
+static const Scope* const& scopeForIdQuery(Context* context, ID id);
+
+// This query always constructs a scope
+// (don't call it if the scope does not need to exist)
+static const owned<Scope>& constructScopeQuery(Context* context, ID id) {
+  QUERY_BEGIN(constructScopeQuery, context, id);
+
+  Scope* result = new Scope();
+
+  if (id.isEmpty()) {
+    // use empty scope for top-level
+  } else {
+    const uast::ASTNode* ast = parsing::idToAst(context, id);
+    if (ast == nullptr) {
+      assert(false && "could not find ast for id");
+    } else {
+      ID parentId = parsing::idToParentId(context, id);
+      result->parentScope = scopeForIdQuery(context, parentId);
+      result->tag = ast->tag();
+      result->id = id;
+      gatherDeclsWithin(ast, result->declared, result->containsUseImport);
+    }
+  }
+  return QUERY_END(toOwned(result));
+}
+
+static const Scope* const& scopeForIdQuery(Context* context, ID id) {
+  QUERY_BEGIN(scopeForIdQuery, context, id);
+
+  const Scope* result = nullptr;
+
+  if (id.isEmpty()) {
+    // use empty scope for top-level
+    const owned<Scope>& newScope = constructScopeQuery(context, id);
+    result = newScope.get();
+  } else {
+    // decide whether or not to create a new scope
+    bool newScope = false;
+
+    const uast::ASTNode* ast = parsing::idToAst(context, id);
+    if (ast == nullptr) {
+      assert(false && "could not find ast for id");
+    } else if (createsScope(ast->tag())) {
+      if (Builder::astTagIndicatesNewIdScope(ast->tag())) {
+        // always create a new scope for a Function etc
+        newScope = true;
+      } else {
+        DeclMap declared;
+        bool containsUseImport = false;
+        gatherDeclsWithin(ast, declared, containsUseImport);
+
+        // create a new scope if we found any decls/uses immediately in it
+        newScope = !(declared.empty() && containsUseImport == false);
+      }
+    }
+
+    if (newScope) {
+      // Construct the new scope.
+      const owned<Scope>& newScope = constructScopeQuery(context, id);
+      result = newScope.get();
+    } else {
+      // find the scope for the parent node and return that.
+      ID parentId = parsing::idToParentId(context, id);
+      result = scopeForIdQuery(context, parentId);
+    }
+  }
+
+  return QUERY_END(result);
+}
+
+const Scope* scopeForId(Context* context, ID id) {
+  return scopeForIdQuery(context, id);
+}
+
+enum UseImportOther {
+  UIO_USE,    // the expr is the thing being use'd e.g. use A.B
+  UIO_IMPORT, // the expr is the thing being imported e.g. import C.D
+  UIO_OTHER   // the expr is something else
```

Comment:
I went "what's other?". I'd suggest `UIO_NEITHER`

Reply:
I don't understand why "neither" would be better than "other" here? They seem equally clear to me. Of course we could name it `UIO_NOT_IMPORT_OR_USE` but that's part of why I don't like adding enums :)

Reply:
I think I am understanding things a bit better, but my thinking was `UIO_OTHER` was representing something other than `use` and `import` that can change visibility. And I didn't know what that was. Then I realized that it represents something **other** than `use` and `import` (i.e something that has nothing to do with visibility).

In concert with the above discussion maybe this enum should be

```chapel
enum VisibilityStmtKind {
  VIS_USE,
  VIS_IMPORT,
  VIS_[NONE/NEITHER]
}
```

That being said, I am not motivated to dwell more on this point, and you can make the final call (including keeping things as-is)

Reply:
renamed